[REVIEW] Use num internal streams instead of creating cumlHandle's inside the C++ layer

teju85 opened the pull request:
This PR is to showcase a possible solution for the issue #931.
However, for this to happen, the constructor for `cumlHandle_impl` had to be updated to expose num-streams parameter.

Tagging @cjnolet @JohnZed and @vishalmehta1991 for review.

teju85 commented:
Folks, the set of changes so far will for sure break the python world. Will fix that soon.

vishalmehta1991 commented on cpp/src/cuML.hpp:
```diff
@@ -33,16 +33,21 @@ class cumlHandle_impl;
 class cumlHandle {
+ private:
+  //TODO: What is the right number?
+  static constexpr int _default_num_internal_streams = 3;
```
I would just keep this one. am not sure how the number 3 was conjured in first place. Do we have a reason for '3' ? 
By default everything has to be in the cuml_handle stream but if more than one stream is needed, user can ask for that either in constructor or at a later stage in algorithm.

teju85 replied:
Good catch. This number was just borrowed from the initial version that Jiri had introduced. :)

I'd probably say even better to change this to 0. So, for any *extra* streams required they need to be explicitly queried. What say?

@cjnolet @dantegd and @JohnZed any objections?

vishalmehta1991 replied:
yes exactly. I agree. extra streams need to be explicit.

cjnolet replied:
0 seems to be a reasonable default, since this is enabling a special case where algorithms can take advantage of multiple stream internally.

teju85 replied:
Done.

vishalmehta1991 commented on cpp/src/randomforest/randomforest_impl.cuh:
```diff
@@ -189,53 +189,56 @@ void rfClassifier<T>::fit(const cumlHandle& user_handle, const T* input,
-  for (int i = 0; i < n_streams; i++) {
-    CUDA_CHECK(cudaStreamCreate(&local_stream[i]));
-    local_handle[i].setStream(local_stream[i]);
+  if (n_streams > handle.getNumInternalStreams()) {
```
this is not very good for performance. I think cuml use allow algorithms to set the number of internal streams .

teju85 replied:
This is a tricky one!

Ideally, we want all resources managed inside `cumlHandle` to have been allocated upfront. That way, it is easier for us to amortize their cost over the run. Now, if we expose a `setNumInternalStreams` method it will break this assumption. This also means that we'll have to sync across all the previous set of streams, before creating these new ones, which in itself can be of some overhead. This is the reason I updated the constructor of `cumlHandle` to accept this parameter during the initialization itself.

In this specific case, instead of printing warning sign and then clamping down the `n_streams` variable, we could throw as `ASSERT`, explicitly asking callers to initialize the `cumlHandle` with appropriate number of internal streams. What say @cjnolet @vishalmehta1991 ?

teju85 replied:
In other words, something like this:
```cpp
const int nStreams = 8;
cumlHandle handle(nStreams);
....
RF_params params;
params.n_streams = nStreams;
...
fit(..., params);
```

vishalmehta1991 replied:
Yeah makes sense. Assert is better for mismatch streams.

teju85 replied:
Thanks. Done.

cjnolet commented on python/cuml/common/handle.pyx:
```diff
@@ -60,8 +60,8 @@ cdef class Handle:
 
-    def __cinit__(self):
-        self.h = <size_t>(new cumlHandle())
+    def __cinit__(self, n_streams=3):
```
Wouldn't we want to set this to the default? I think RF is going to be the only algorithm using this currently, perhaps KNN in the near future.

teju85 replied:
Done.

cjnolet commented on python/cuml/metrics/trustworthiness.pyx:
```diff
@@ -102,7 +102,7 @@ def trustworthiness(X, X_embedded, handle=None, n_neighbors=5,
     cdef cumlHandle* handle_ = <cumlHandle*>0
     if handle is None:
-        handle_ = <cumlHandle*><size_t>(new cumlHandle())
+        handle_ = <cumlHandle*><size_t>(new cumlHandle(3))
```
Wouldn't we want to set this to the default as well? Trustworthiness isn't making use of this property.

teju85 replied:
Done.

cjnolet commented on wiki/cpp/DEVELOPER_GUIDE.md:
````diff
@@ -259,6 +259,8 @@ void foo(const ML::cumlHandle_impl& h, ...)
 }
 ```
 
+An example of how to use internal streams to schedule work on a single GPU can be found in [here](https://github.com/rapidsai/cuml/pull/1015). This PR uses the internal streams inside `cumlHandle_impl` to schedule more work onto the GPU for Random Forest building.
````
We should follow the format we've been using for the rest of the developer guide and provide the example in place. What do you think?

teju85 replied:
Fair point. Done. Can you check now?

teju85 commented:
I think I have addressed all the review comments. @vishalmehta1991 and @cjnolet please check now.

Also, @vishalmehta1991 had concerns about the conflicts arising due to this PR and his PR #961 . We should discuss how to resolve this before merging either of the two.

teju85 commented:
IMO, it is better to take the changes in #961 first, followed by me resolving conflicts arising with the current PR.

dantegd commented:
@teju85 added the `Blocked!` label just to reflect that this PR is waiting on 961

teju85 commented:
having the same issue as the other PR #823.

teju85 commented:
rerun tests

cjnolet commented:
rerun tests

cjnolet commented:
rerun tests

teju85 commented:
JFYI, @vishalmehta1991 has requested to stop merging this PR until the PR #1087 gets through.

teju85 commented:
@dantegd any ideas why I get the following error in CI?
`E   ImportError: cannot import name 'TOTAL_MEMORY' from 'distributed.worker' (/conda/envs/gdf/lib/python3.7/site-packages/distributed/worker.py)`

cjnolet commented on wiki/cpp/DEVELOPER_GUIDE.md:
```diff
@@ -6,6 +6,7 @@ Please start by reading [CONTRIBUTING.md](../../CONTRIBUTING.md).
 
 ## Performance
 1. In performance critical sections of the code, favor `cudaDeviceGetAttribute` over `cudaDeviceGetProperties`. See PR [#973](https://github.com/rapidsai/cuml/pull/973) for more details.
+2. If an algo requires you to launch GPU work in multiple cuda streams, do not create multiple `cumlHandle` objects, one for each such work stream. Instead, expose a `n_streams` parameter in that algo's cuML C++ interface and then rely on `cumlHandle_impl::getInternalStream()` to pick up the right cuda stream. See PR [#1015](https://github.com/rapidsai/cuml/pull/1015) and also the section on [CUD ...
```
I'm not sure how I missed this. I'd prefer not to point users to pull requests in the developer guide as it's not straightforward and can quickly get out of date as code is updated.

teju85 replied:
What do you recommend instead?

cjnolet replied:
IMO, the link to CUDA Resources and the TIP are good enough. Maybe we could also link to the example in the `threading` section. What do you think?

teju85 replied:
Done. How about now?

cjnolet commented on wiki/cpp/DEVELOPER_GUIDE.md:
````diff
@@ -259,6 +260,16 @@ void foo(const ML::cumlHandle_impl& h, ...)
 }
 ```
 
+The example below shows one way to create `nStreams` number of internal cuda streams which can later be used by the algos inside cuML. For a full working example of how to use internal streams to schedule work on a single GPU, the reader is further referred to [this PR](https://github.com/rapidsai/cuml/pull/1015). In this PR, the internal streams inside `cumlHandle_impl` are used to schedule more ...
````
I plan to use this example in the threading update PR as well, as I've been waiting for this PR to get merged so that I can update it. IIRC, the RF PR specifically uses OMP to schedule the streams in multiple threads and I've captured an example of this in the threading updates to the developer guide already. 

Since this is a section on threading, I will probably combine the example with the one in that PR and also incorporate the more precise wording that Jiri & I have been working on in that PR.

teju85 replied:
Makes sense. Thanks

teju85 commented:
rerun tests

teju85 commented:
rerun tests

teju85 commented:
@dantegd @cjnolet Any suggestions on how to fix this CI error? `E   TypeError: no default __reduce__ due to non-trivial __cinit__`

I need to have a non-default ctor in handle.pyx for the python users to be able to specify the number of streams to be created inside `cumlHandle`.

cjnolet commented:
@teju85, I see what's going on here. The problem is not because you have a non-default `__cinit__()`, it's because the Dask RF code is trying to pickle the  handle to send it to the workers, and pickling a cython class with a non-default `__cinit__()` requires a non-default `__reduce()__` (because there's a well-defined separation between cython variables, which aren't natively pickleable, and Python objects, which are natively PIckleable. 

Here's the code that's giving the problem:

```
        if handle is None:
            handle = cuml.Handle(n_streams)

        self.rfs = {
            worker: c.submit(
                RandomForestClassifier._func_build_rf,
                n,
                self.n_estimators_per_worker[n],
                max_depth,
                handle,
                max_features,
                n_bins,
                split_algo,
                split_criterion,
                min_rows_per_node,
                bootstrap,
                bootstrap_features,
                type_model,
                verbose,
                rows_sample,
                max_leaves,
                n_streams,
                quantile_per_tree,
                dtype,
                random.random(),
                workers=[worker],
            )
            for n, worker in enumerate(workers)
        }
```

The fix is to pass `n_streams` to the workers and have the `RandomForestClassifier._func_build_rf` create the handle (locally on each worker). I'll fix this for you and push, since we're strapped for time in 0.10.

cjnolet commented:
If we want to enable the sharing of a cumlHandle on the workers across different runs of algorithms (which will get tricky using NCCL in the comms), we will want to cache the handle on the workers (look into `CommsContext` to see how I'm doing this). Problem is, this won't be thread-safe, so it might be worth caching based on the id of the thread, and perhaps using some sort of LRU strategy.

teju85 commented:
Awesome. Thanks @cjnolet for getting this PR finally across the border!
